### PR TITLE
Escape usernames in the frontend

### PR DESCRIPTION
**Pull request by nicorikken**
To cope with special characters like backslash, to address issue
https://github.com/jupyterhub/jupyterhub/issues/2128

### CODE REVIEW

**Comment by nicorikken** [comment]
**I haven't check its workings yet, this is based on assumptions. Please test.**

**Comment by nicorikken** [comment]
I've come to the realization that in regards to spawning, this issue is probably deeper, as the username is not properly escaped in the python handlers and the HTML templating.

**Comment by nicorikken** [comment]
I was mistaken. `window.jhdata.user` doesn't include the backlash. So this makes me question where the backslash error is happening.

**Comment by minrk** [comment]
I think your final code is right, that this should be using encodeURIComponent instead of escape. Can you rebase to remove the Python-side changes that are now unused, and just the js fixes? Then I'm happy to merge.

File: jupyterhub/user.py [hunk 1]
```diff
@@ -350,6 +351,11 @@ def escaped_name(self):
         """My name, escaped for use in URLs, cookies, etc."""
         return quote(self.name, safe='@~')
 
+    @property
+    def unicode_escaped_name(self):
+        """My name, escaped for use in javascript inserts, etc."""
```
Comment (willingc) [inline jupyterhub/user.py]:
```suggestion
        """The user name, escaped for use in javascript inserts, etc."""
```

File: share/jupyterhub/static/js/home.js [hunk 2]
```diff
@@ -103,7 +103,7 @@ require(["jquery", "moment", "jhapi", "utils"], function(
   $(".new-server-btn").click(function() {
     var row = getRow($(this));
     var serverName = row.find(".new-server-name").val();
-    window.location.href = "../spawn/" + user + "/" + serverName;
+    window.location.href = "../spawn/" + escape(user) + "/" + serverName;
```
Comment (willingc) [inline share/jupyterhub/static/js/home.js]: @minrk did you want this to be encodeURIComponent instead of escape?

**Comment by nicorikken** [comment]
Right... that was a lot of back-and-forth. @willingc @minrk thanks for responding so quickly, I didn't notice while I was still fixing this code. Now it is neatly rounded. The main issue was that the Python string needed to be properly JSON-encoded. My previous version using unicode encoding didn't cover all cases, so I went this route.

- I chose a substring `[1:-1]` to strip the outer JSON quotes, to remain compatible with the `user.name` and `user.escaped_name` properties. The outer quotes are added in the templating.
- The parameter is renamed to `json_escaped_name` to better reflect its purpose.
- A few tests are added for edge-cases, although it seems to break the build somehow.

File: share/jupyterhub/static/js/home.js (hunk 2 above)
Comment (nicorikken) [reply share/jupyterhub/static/js/home.js]: My bad, this introduced errors, it was a bad attempt.

File: jupyterhub/user.py (hunk 1 above)
Comment (nicorikken) [reply jupyterhub/user.py]: Thanks, resolved.

**Comment by nicorikken** [comment]
Done! Ready to merge now.
It was nice to work on JupyterHub and get to know its internals.

**Comment by nicorikken** [comment]
By the way, I tested this on our own infrastructure, where the issue occurred as well. This PR solved that.

**Comment by minrk** [comment]
That works, thanks!
